File: blockchain/p2p-blockchain backend/blockchain.py
import hashlib
import time
import json

class Transaction:
    def __init__(self, sender, recipient, amount):
        self.sender = sender
        self.recipient = recipient
        self.amount = amount
        self.timestamp = time.time()

    def to_dict(self):
        return {
            'sender': self.sender,
            'recipient': self.recipient,
            'amount': self.amount,
            'timestamp': self.timestamp
        }

class Block:
    def __init__(self, index, transactions, timestamp, previous_hash):
        self.index = index
        self.transactions = transactions
        self.timestamp = timestamp
        self.previous_hash = previous_hash
        self.nonce = 0
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        block_string = json.dumps({
            'index': self.index,
            'transactions': [tx.to_dict() for tx in self.transactions],
            'timestamp': self.timestamp,
            'previous_hash': self.previous_hash,
            'nonce': self.nonce
        }, sort_keys=True)
        return hashlib.sha256(block_string.encode()).hexdigest()

    def mine_block(self, difficulty):
        while self.hash[:difficulty] != '0' * difficulty:
            self.nonce += 1
            self.hash = self.calculate_hash()

class Blockchain:
    def __init__(self):
        self.chain = [self.create_genesis_block()]
        self.difficulty = 4
        self.pending_transactions = []
        self.nodes = set()

    def create_genesis_block(self):
        return Block(0, [], time.time(), "0")

    def get_latest_block(self):
        return self.chain[-1]

    def mine_block(self):
        block = Block(len(self.chain), self.pending_transactions, time.time(), self.get_latest_block().hash)
        block.mine_block(self.difficulty)
        self.chain.append(block)
        self.pending_transactions = []
        return block

    def add_transaction(self, sender, recipient, amount):
        self.pending_transactions.append(Transaction(sender, recipient, amount))

    def is_chain_valid(self):
        for i in range(1, len(self.chain)):
            current = self.chain[i]
            previous = self.chain[i-1]
            if current.hash != current.calculate_hash():
                return False
            if current.previous_hash != previous.hash:
                return False
            if current.hash[:self.difficulty] != '0' * self.difficulty:
                return False
        return True
     

File: project 2/src/blockchain/Blockchain.ts
import { Block, Transaction } from '../types/blockchain';

export class Blockchain {
  private chain: Block[] = [];
  private difficulty: number = 2;

  constructor() {
    this.chain = [this.createGenesisBlock()];
  }

  private createGenesisBlock(): Block {
    return {
      index: 0,
      timestamp: Date.now(),
      transactions: [],
      previousHash: '0',
      hash: this.calculateHash(0, Date.now(), [], '0', 0),
      nonce: 0,
      miner: 'genesis',
      difficulty: this.difficulty
    };
  }

  private calculateHash(index: number, timestamp: number, transactions: Transaction[], previousHash: string, nonce: number): string {
    const data = `${index}${timestamp}${JSON.stringify(transactions)}${previousHash}${nonce}`;
    return this.sha256(data);
  }

  private sha256(data: string): string {
    let hash = 0;
    for (let i = 0; i < data.length; i++) {
      const char = data.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16).padStart(8, '0');
  }

  getLatestBlock(): Block {
    return this.chain[this.chain.length - 1];
  }

  async mineBlock(transactions: Transaction[], minerAddress: string): Promise<Block> {
    const previousBlock = this.getLatestBlock();
    const newBlock: Block = {
      index: previousBlock.index + 1,
      timestamp: Date.now(),
      transactions: [...transactions],
      previousHash: previousBlock.hash,
      hash: '',
      nonce: 0,
      miner: minerAddress,
      difficulty: this.difficulty
    };

    const target = '0'.repeat(this.difficulty);

    while (true) {
      newBlock.hash = this.calculateHash(
        newBlock.index,
        newBlock.timestamp,
        newBlock.transactions,
        newBlock.previousHash,
        newBlock.nonce
      );

      if (newBlock.hash.startsWith(target)) {
        break;
      }

      newBlock.nonce++;

      if (newBlock.nonce % 1000 === 0) {
        await new Promise(resolve => setTimeout(resolve, 0));
      }
    }

    return newBlock;
  }

  addBlock(block: Block): boolean {
    if (this.isValidBlock(block)) {
      this.chain.push(block);
      return true;
    }
    return false;
  }

  private isValidBlock(block: Block): boolean {
    const previousBlock = this.getLatestBlock();
    
    if (block.index !== previousBlock.index + 1) return false;
    if (block.previousHash !== previousBlock.hash) return false;
    
    const calculatedHash = this.calculateHash(
      block.index,
      block.timestamp,
      block.transactions,
      block.previousHash,
      block.nonce
    );
    
    return block.hash === calculatedHash && block.hash.startsWith('0'.repeat(this.difficulty));
  }

  getChain(): Block[] {
    return [...this.chain];
  }

  replaceChain(newChain: Block[]): boolean {
    if (newChain.length > this.chain.length) {
      this.chain = [...newChain];
      return true;
    }
    return false;
  }

  getDifficulty(): number {
    return this.difficulty;
  }
}